### DISCUSSION TITLE
Dual carriageway and left turn, Continue instruction

### DISCUSSION THREAD

**Discussion by r11d11** [devemux86/cruiser#307] [Q&A] [answered]
Hello, last week i drove 1000km with my motorbike and I had a lot of fun with Cruiser, Mapsforge and Brouter. Best combination ever.
One small issue.
When I had to turn left on a dual carriageway Cruiser says I have to go straight and wenn I have reached the  central reservation of the street it says I have to go left. These two commands come within milliseconds and you have no chance to turn left if you only look at the the arrow top left.

I would like it better if Cruiser would suppress the first indication straight ahead when Cruiser notices that you have to turn left 20 metres further on. Or is there an option in Cruiser that 2 commands are combined

An example: 
[four lane route and left turn.zip](https://github.com/devemux86/cruiser/files/12565241/four.lane.route.and.left.turn.zip)

![turn_left_corr](https://github.com/devemux86/cruiser/assets/101214119/86244a48-68d9-4162-9382-123160b4613f)

**Comment by devemux86** [comment]
@r11d11 thanks for using Cruiser and the report!

With which router was the route calculated?

Cruiser currently does not change the turn instructions, it uses them as they come from the router.

Αs you mentioned, maybe it's better to merge such close "continue" instructions to avoid confusion.
I have already put it in my implementation list.

---

BRouter also seems to have plans to post-process such turn instructions:
- https://github.com/abrensch/brouter/issues/613

**Reply by r11d11** [reply]
It was calculated with Brouter offline and profil Auto. 

>Cruiser currently does not change the turn instructions, it uses them as they come from the router.
Then I have to write this issue in the Brouter forum if a Brouter forum exists

**Reply by devemux86** [reply]
> It was calculated with Brouter offline and profil Auto.
> Then I have to write this issue in the Brouter forum if a Brouter forum exists

@r11d11 Please also mention your example in the BRouter issue, better as a [BRouter website](https://brouter.de/brouter-web/) route link:

- https://github.com/abrensch/brouter/issues/613

cc @afischerdev

**Reply by r11d11** [reply]
I now wrote this issue in the brouter forum. Thank you for your advice and your very good work

**Comment by devemux86** [comment]
@r11d11 

In your image, the 2nd instruction (430 m) is not "Turn left", but arrive at destination?

**Reply by r11d11** [reply]
That is true. That is a bit strange.
But there a 2 turning instruction. "Links abbiegen" means turn left

![Turnlef1t](https://github.com/devemux86/cruiser/assets/101214119/94191a27-d5d6-44de-be00-514cea21a8cd)

**Reply by devemux86** [reply]
Loading your route, I get the following navigation:

<img src="https://github.com/devemux86/cruiser/assets/3484020/3b58c60b-fe5f-4c97-b322-369cc4e74ad4" width="50%"/>

**Reply by r11d11** [reply]
I simulated it again an now I get the same left turn hint as you. I can not explain why it is missing in my first screenshot. The screenshot is older. Maybe it was taken with an older version of cruiser? With 3.0.31 it now looks for me like your screenshot. Sorry for the confusion

**Reply by devemux86** [reply]
@r11d11 probably that route did not have (?) the "turn left" instruction.

Can you please update the screenshot in the BRouter issue, so its readers can see the correct navigation?
- https://github.com/abrensch/brouter/issues/625

**Reply by r11d11** [reply]
i now updated my first screenshot

**Comment by Ciutnik** [comment]
It is true that this behavior is very misleading. These have already been discussed here: https://github.com/devemux86/cruiser/discussions/164#discussioncomment-6730178

@devemux86 Maybe this whole part of the discussion should be moved to the current topic? It will be easier to find it in the future here.

**Reply by devemux86** [reply]
> Maybe this whole part of the discussion should be moved to the current topic?

Unfortunately GitHub does not seem to offer such function.

**Comment by Ciutnik** [comment]
However, something is not as it should be. Left turn instructions are not displayed here:

<img src="https://github.com/devemux86/cruiser/assets/122647782/e3ed2c6c-ee16-4274-8a43-d6471ab8bb09" width="50%"/>

This is the link to this maneuver on the BRouter website: [Left turn](https://brouter.de/brouter-web/#map=18/52.22868/21.08535/standard&lonlats=21.085318,52.228593;21.086476,52.22807&profile=car-eco)

**Reply by devemux86** [reply]
@Ciutnik thanks for the report.

Probably you should also mention it in BRouter repository:
- https://github.com/abrensch/brouter/issues/625

Attached is the route response, it does not contain a left turn:
[Warsaw 0.1km.gpx.txt](https://github.com/devemux86/cruiser/files/12671819/Warsaw.0.1km.gpx.txt)

To export it, follow the instructions here:
- https://github.com/devemux86/cruiser/discussions/250#discussioncomment-6508754

**Reply by Ciutnik** [reply]
That's right. Also the Cruiser export does not include a left turn, which is understandable.
[Left_turn.gpx.txt](https://github.com/devemux86/cruiser/files/12671880/Left_turn.gpx.txt)

> Probably you should also mention it in BRouter repository

Done!

**Comment by devemux86** [comment]
Cruiser already processes the turn instructions of all routing services to prepare them for navigation.
I will add another process step that checks the "Continue" instructions if they are close to other turns.

We need a distance threshold, e.g. 50 m. Fixed or in settings?
It should probably be different for driving / cycling / walking?

**Reply by devemux86** [reply]
We can use as threshold one of the distances that can already be adjusted in settings.
e.g. "Last announcement" or "Small distance"

The default distances are shown below for driving / cycling / walking:

<img src="https://github.com/devemux86/cruiser/assets/3484020/33441d1e-8a1d-43cb-8b74-e420f3af29b7" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/9f160116-cc4f-4dfa-a632-e1017c568925" width="33%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/21fdd37c-d898-4dad-9f97-659bc595bac3" width="33%"/>

**Reply by devemux86** [reply]
Post process of turn instructions (using the "Last announcement" distance) is available in the latest:
- https://github.com/devemux86/cruiser/discussions/390

Please test your routes to see if everything works correctly. :slightly_smiling_face:

**Reply by Ciutnik** [reply]
Great! I checked: I won't get lost at nearby dual carriageway junctions due to confusing instructions! Well done 😀 I'm really happy about this. Thank you very much! 👍

**Reply by devemux86** [reply]
@r11d11 _sorry for accidentally deleting your post, copying from the email below:_

Nice that you are still working on the topic. I thought that brouter has to change something.
Dowlaoded b10 and did 3 testing with b10.
Routes were calculated with Brouter offline and profil Auto.
The first route I tested with left turn and dual carriageway junctions now works better with b10:
[ok.kurviger.txt](https://github.com/devemux86/cruiser/files/13619879/ok.kurviger.txt)
Very nice.
But the following 2 test are even worse in b10, because at the left turn now only a straight ahead arrow is displayed.
notworking1 is the example in my first post from 9. Sept. with the picture.
[notworking1.kurviger.txt](https://github.com/devemux86/cruiser/files/13619906/notworking1.kurviger.txt)
Same issue here:
[notworking2.kurviger.txt](https://github.com/devemux86/cruiser/files/13619908/notworking2.kurviger.txt)
Unfortunately, the wrong turning point has been deleted

_Originally posted by @r11d11._

**Reply by devemux86** [reply]
@r11d11 thanks for the examples.

Note: only the "continue" instructions may be removed and if are **near** other instructions.

In your last 2 examples BRouter (offline) "original" instructions do NOT contain "Turn left".
They only contain the "Continue" instruction, so Cruiser has not anything to post process.

I also mentioned them in the BRouter repository, they already work on this:
- https://github.com/abrensch/brouter/issues/613#issuecomment-1750776070

cc @afischerdev

**Comment by devemux86** [comment]
As a reminder, there is work in progress in BRouter's turn instructions:
- https://github.com/abrensch/brouter/issues/613

cc @afischerdev

**Comment by devemux86** [comment, marked as answer]
Post process of turn instructions to remove nearby "Continue" is available in [Cruiser 3.0.37](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
